### ISSUE TITLE
remove mixing down to mono to allow stereo input device transmission

### ISSUE THREAD

**Issue by k1ttt-remote** [mumble-voip/mumble#4610]
It is becoming quite common in amateur radio for operators to listen to 2 radios at once, one in each ear.  Mumble has become popular for streaming radio audio because of it's low latency but is limited to mixing the 2 radio streams.  Instead of using a bot to feed stereo music as has already been done in #4209 I and apparently several others would like to be able to feed stereo audio inputs like a line-in or stereo-mix device in without it down mixing to mono for transmission.  

In #3098 a hidden option to set a mask was added to allow selecting which stereo input channels get downmixed to mono for transmission so it sounds like the input device audio may already be available in stereo, and since #4209 covers transmitting and receiving stereo, how about removing the downmix to mono to let both channels go to opus for transmission?

i think i have the 1.4.0 build environment set up, can someone tell me where in the audio input processing this downmixing occurs so i can try bypassing it?  Since the receiving client should be able to handle the stereo audio stream from #4209 and the sending client can already send stereo from other sources this should be relatively minimal change, but as others have noted the audio code is rather convoluted right now.

**Comment by Krzmbrzl** [comment]
Is this perhaps related to (or even a duplicate of) #4112 ?

> can someone tell me where in the audio input processing this downmixing occurs so i can try bypassing it?

I assume that this happens somewhere in the `AudioInput` class or maybe even in the backend-specifc code (e.g. `WASAPI.cpp`). @davidebeatrici  can probably give you a better hint though...

**Comment by k1ttt-remote** [comment]
#4112 sounds more related to mixing at the audio output and doesn’t take into account the work already done to use stereo sound bot transmission and reception.  My #4610 expands on my comment near the end of #4112 for audio input.

 

David Robbins K1TTT
e-mail:  <mailto:k1ttt@arrl.net> mailto:k1ttt@arrl.net
web:  <http://wiki.k1ttt.net/> http://wiki.k1ttt.net
AR-Cluster node:  <telnet://k1ttt.net:7373/> telnet://k1ttt.net:7373

**Comment by k1ttt-remote** [comment]
Perhaps I should have asked this first… if I succeed in getting mumble 1.4.0 to send stereo audio will it pass through a murmer 1.3.3 server and be properly reproduced by a mumble 1.3.3 client?

 

David Robbins K1TTT
e-mail:  <mailto:k1ttt@arrl.net> mailto:k1ttt@arrl.net
web:  <http://wiki.k1ttt.net/> http://wiki.k1ttt.net
AR-Cluster node:  <telnet://k1ttt.net:7373/> telnet://k1ttt.net:7373

**Comment by Krzmbrzl** [comment]
> if I succeed in getting mumble 1.4.0 to send stereo audio will it pass through a murmer 1.3.3 server and be properly reproduced by a mumble 1.3.3 client?

It will definitely pass through the server (it only re-distributes the audio data it received without looking at it). I believe though that any older client will mix the audio down to mono again (the ability to play stereo is also a new feature in 1.4.0)

**Comment by k1ttt-remote** [comment]
Hmmm, is there an installer for 1.4.0 with stereo available anywhere yet?    Is it easier to set up the build environment under azure than vs2019 on my desktop??

 

David Robbins K1TTT
e-mail:  <mailto:k1ttt@arrl.net> mailto:k1ttt@arrl.net
web:  <http://wiki.k1ttt.net/> http://wiki.k1ttt.net
AR-Cluster node:  <telnet://k1ttt.net:7373/> telnet://k1ttt.net:7373

**Comment by Krzmbrzl** [comment]
> Hmmm, is there an installer for 1.4.0 with stereo available anywhere yet?

We are working on releasing the first snapshots releases (hopefully) in the coming weeks. Until then there's always the option of grabbing an installer from our AzurePipelines CI jobs.

> Is it easier to set up the build environment under azure than vs2019 on my desktop??

I have no idea what you mean by that xD
And I'm also not a Windows-guy (doing basically all development on Linux here).
What I can say with certainty though is that with 1.4.0 building the environment has become much, much easier thanks to the use of vcpkg. If you want to give it a try, have a look at https://github.com/mumble-voip/mumble/blob/master/docs/dev/build-instructions/build_static.md

**Comment by TerryGeng** [comment]
Actually, there were some discussions about feeding stereo stream into Mumble client itself some time ago in the IRC channel (perhaps in some issues here as well).

I think at that time the conclusion reached was Mumble client is not designed to handle stereo input. Clearly, it was designed to handle the human voice (with all RNN and echo cancellation and opus set to voice mode), and stereo things (in your case, two radio signals in each ear) are clearly... not so human.

The conclusion was: use a bot for the stereo thing. pymumble and https://github.com/ranomier/pymumble-abot could be possible solution. People complained about the maintenance of https://github.com/ranomier/pymumble-abot is stalled. I'm still wondering if I can take the role of maintainer if necessary.

TL;DR: As for your special use case, I would suggest you start two mumble client instance and use the manual position plugin to position each of these two clients to total left and right.

**Comment by k1ttt-remote** [comment]
I thought about 2 instances, but decided that would be too confusing to set up and would be susceptible to the two streams getting out of sync which would be annoying when switching back and forth.

 

I also tried hacking the mumble code to get stereo through it without success.  Opus has some settings that make you think it will do stereo, but while I was able to get the stereo streams past the mumble multiple stream downsamplers, I could not get opus to actually encode it… apparently at lower data rates or bandwidth settings like would be used for just voice opus will combine the channels on its own.

 

So I am now working on my own app that is greatly simplified from mumble which separates the left and right, sends each through it’s own codec, puts them into a single udp transmission and then reverses the process at the receiving end.   

 

David Robbins K1TTT
e-mail:  <mailto:k1ttt@arrl.net> mailto:k1ttt@arrl.net
web:  <http://wiki.k1ttt.net/> http://wiki.k1ttt.net
AR-Cluster node:  <telnet://k1ttt.net:7373/> telnet://k1ttt.net:7373

**Comment by TerryGeng** [comment]
In this case, I suggest you hack  [about](https://github.com/ranomier/pymumble-abot) instead because it is much easier to work on. To hack mumble, you would walk into the swamp of huge tons of features that are irrelevant to your need.

Creating your own app is also a choice if you don't need the other features mumble supports like ACLs, groups, text messages, etc.

I think if there isn't anything else to discuss, I think we can close this issue since it duplicates #4112 (or it belongs to part of #4112). @Krzmbrzl

**Comment by Krzmbrzl** [comment]
>  I think we can close this issue since it duplicates #4112 (or it belongs to part of #4112). 

I agree. If it turns out that this is indeed an issue that should exist in and of its own, then I'll reopen it again :point_up:
